docs/rfc/rfc-002-ipc-ecosystem.md

```diff
@@ -0,0 +1,348 @@
+blocked, which is good for security but complicates use by the light client.
+
+TODO: Figure out what is going on with the weird pseudo-inheritance service
+interface.  The BaseService type makes sense, but its embeds are more
+complicated.
+
+### Context: Remote Signer Issues
+
+Since the remote signer needs a secure communication channel to exchange keys
+and signatures, and is expected to run truly remotely from the node (i.e., on a
+separate physical server), there is not a whole lot we can do here. We should
+finish the deprecation and removal of the "raw" socket protocol between the
+consensus node and remote signers, but the use of gRPC is appropriate.
+
+The main improvement we can make is to simplify the implementation quite a bit,
+once we no longer need to support both "raw" and gRPC transports.
+
+### Context: ABCI Issues
+
+In the original design of ABCI, the presumption was that all access to the
+application should be mediated by the consensus node. The idea is that outside
+access could change application state and corrupt the consensus process, which
+relies on the application to be deterministic. Of course, even without outside
+access an application could behave nondeterministically, but allowing other
+programs to send it requests was seen as courting trouble.
+
+Conversely, users noted that most of the time, tools written for a particular
+application don't want to talk to the consensus module directly. The
+application "owns" the state machine the consensus engine is replicating, so
+tools that care about application state should talk to the application.
+Otherwise, they would have to bake in knowledge about Tendermint (e.g., its
+interfaces and data structures) just because of the mediation.
+
+For clients to talk directly to the application, however, there is another
+concern: The consensus node is the ABCI _client_, so it is inconvenient for the
+application to "push" work into the consensus module via ABCI itself.  The
+current implementation works around this by calling the consensus node's RPC
+service, which exposes an `ABCIQuery` kitchen-sink method that allows the
+application a way to poke ABCI messages in the other direction.
+
+Without this RPC method, you could work around this (at least in principle) by
+having the consensus module "poll" the application for work that needs done,
+but that has unsatisfactory implications for performance and robustness, as
+well as being harder to understand.
+
+There has apparently been discussion about trying to make a more bidirectional
+communication between the consensus node and the application, but this issue
+seems to still be unresolved.
+
+Another complication of ABCI is that it requires the application (server) to
+maintain [four separate connections][abci-conn]: One for "consensus" operations
+(BeginBlock, EndBlock, DeliverTx, Commit), one for "mempool" operations, one
+for "query" operations, and one for "snapshot" (state synchronization) operations.
```

Comment:
Is there a lock for each connection or across all four connections? If I recall correctly, I remember raising a few eyebrows to the way things constructed (in regards to the ABCI server/client stuff)

Reply:
I don't have a clear answer to this question yet. As far as I can tell there are _at least_ two exclusion locks for the protocol itself, one for consensus operations and one for application state. The clients also have their own locks to coordinate a somewhat confusing interaction between synchronous and asynchronous calls. But that's just the explicit locks: There is also a network of channels for exchanging requests and responses, and a channel is essentially also a shared lock. That's in addition to any synchronization the SDK is doing internally, which I haven't really traced out yet.  

The fact that it's so hard to understand how and where synchronization is being done is the root of the problem. I think we can make that easier to understand (and possibly perform better) without having to compromise the functionality of the protocol.

Reply:
As a data point, the spec does list that there is a lock across _all_ connections. See [the concurrency section of the ABCI spec](https://github.com/tendermint/spec/blob/master/spec/abci/apps.md#concurrency). I think we should verify that this is still the case, but it's at least document that way.

Reply:
> As a data point, the spec does list that there is a lock across all connections.

That's right, though just to be clear—that's an implementation choice rather than a spec requirement. There are some ongoing attempts (e.g., #6899) to reduce the locking, but it's tricky because there's no single place where a scheduling decision can be made. It's hard to be sure that an unsynchronized access is actually safe. The tests don't exercise those paths nearly enough for me to feel confident about it.